src/lib/homestore_backend/hs_shard_manager.cpp

```diff
@@ -1,34 +1,51 @@
-#include "homeobject.hpp"
-#include "replication_message.hpp"
-
 #include <homestore/homestore.hpp>
 #include <homestore/blkdata_service.hpp>
 #include <homestore/meta_service.hpp>
 
+#include "hs_homeobject.hpp"
+#include "replication_message.hpp"
+#include "replication_state_machine.hpp"
+#include "lib/homeobject_impl.hpp"
+
 namespace homeobject {
 
+static constexpr uint32_t SEQUENCE_BIT_NUM_IN_SHARD{48};
+
 uint64_t ShardManager::max_shard_size() { return Gi; }
 
-uint64_t ShardManager::max_shard_num_in_pg() { return ((uint64_t)0x01) << shard_width; }
+uint64_t ShardManager::max_shard_num_in_pg() { return ((uint64_t)0x01) << SEQUENCE_BIT_NUM_IN_SHARD; }
+
+shard_id_t HSHomeObject::generate_new_shard_id(pg_id_t pgid) {
+    HS_PG* pg{nullptr};
+    {
+        std::shared_lock lg{_pg_lock};
+        if (auto it = _pg_map.find(pgid); it != _pg_map.end()) {
+            pg = s_cast< HS_PG* >(it->second.get());
+        } else {
+            RELEASE_ASSERT(false, "commit on a shard with unavailable pg [{}]", pgid);
+            return 0;
+        }
+    }
+
+    std::scoped_lock lg(pg->mtx_);
+    auto new_sequence_num = ++(pg->shard_sequence_num_);
+    RELEASE_ASSERT_LT(new_sequence_num, ShardManager::max_shard_num_in_pg(),
+                      "new shard id must be less than ShardManager::max_shard_num_in_pg()");
+    return make_new_shard_id(pgid, new_sequence_num);
+}
 
-shard_id HSHomeObject::generate_new_shard_id(pg_id pg) {
-    std::scoped_lock lock_guard(_pg_lock);
-    auto iter = _pg_map.find(pg);
-    RELEASE_ASSERT(iter != _pg_map.end(), "Missing pg info");
-    auto new_sequence_num = ++(iter->second.shard_sequence_num);
-    RELEASE_ASSERT(new_sequence_num < ShardManager::max_shard_num_in_pg(),
-                   "new shard id must be less than ShardManager::max_shard_num_in_pg()");
-    return make_new_shard_id(pg, new_sequence_num);
+shard_id_t HSHomeObject::make_new_shard_id(pg_id_t pg, uint64_t sequence_num) const {
+    return ((uint64_t)pg << SEQUENCE_BIT_NUM_IN_SHARD) | sequence_num;
```

Comment:
I think a lot of this was removed; maybe added back in a merge? It's shared by memory through homeobject_impl.hpp.

Reply:
Fixed